libfwupd/fwupd-device.c

```diff
@@ -2600,32 +2604,34 @@ fwupd_device_to_string (FwupdDevice *self)
 		fwupd_pad_kv_str (str, FWUPD_RESULT_KEY_STATUS,
 				  fwupd_status_to_string (priv->status));
 	}
-	if (priv->guids->len > 0) {
-		g_autoptr(GHashTable) ids = NULL;
-		ids = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
-		for (guint i = 0; i < priv->instance_ids->len; i++) {
-			const gchar *instance_id = g_ptr_array_index (priv->instance_ids, i);
-			g_hash_table_insert (ids,
-					     fwupd_guid_hash_string (instance_id),
-					     g_strdup (instance_id));
-		}
-		for (guint i = 0; i < priv->guids->len; i++) {
-			const gchar *guid = g_ptr_array_index (priv->guids, i);
-			const gchar *instance_id = g_hash_table_lookup (ids, guid);
-			if (instance_id == NULL) {
-				fwupd_pad_kv_str (str, FWUPD_RESULT_KEY_GUID, guid);
-			} else {
-				g_autofree gchar *tmp = NULL;
-				tmp = g_strdup_printf ("%s <- %s", guid, instance_id);
-				fwupd_pad_kv_str (str, FWUPD_RESULT_KEY_GUID, tmp);
-			}
-		}
-	} else {
-		for (guint i = 0; i < priv->instance_ids->len; i++) {
-			const gchar *instance_id = g_ptr_array_index (priv->instance_ids, i);
-			fwupd_pad_kv_str (str, FWUPD_RESULT_KEY_INSTANCE_IDS, instance_id);
+
+	/* show InstanceIDs optionally mapped to GUIDs, and also "standalone" GUIDs */
+	instance_ids_by_guid = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+	for (guint i = 0; i < priv->instance_ids->len; i++) {
+		const gchar *instance_id = g_ptr_array_index (priv->instance_ids, i);
+		g_hash_table_insert (instance_ids_by_guid,
+				     fwupd_guid_hash_string (instance_id),
+				     g_strdup (instance_id));
+	}
+	g_hash_table_iter_init (&iter, instance_ids_by_guid);
+	while (g_hash_table_iter_next (&iter, &key, &value)) {
+		const gchar *guid = (const gchar *) key;
+		const gchar *instance_id = (const gchar *) value;
+		g_autofree gchar *tmp = NULL;
+		if (fwupd_device_has_guid (self, guid)) {
+			tmp = g_strdup_printf ("%s <- %s", guid, instance_id);
+		} else {
+			tmp = g_strdup_printf ("%s <- %s (!)", guid, instance_id);
 		}
```

Comment:
nit; i don't think you need braces here

Reply:
I did wonder about:

		tmp = g_strdup_printf ("%s <- %s%s", guid, instance_id,
				       fwupd_device_has_guid (self, guid) ? "" : " (!)");

Reply:
maybe just a GString would be better here.

Reply:
Yeah either a GString or that are better.

Reply:
I reworked it a bit to be more efficient too. Looks better?